src/main/java/com/commercetools/sync/customers/helpers/CustomerReferenceResolver.java

```diff
@@ -0,0 +1,167 @@
+     * @return a {@link CompletionStage} that contains as a result a new builder instance with resolved customer group
+     *         reference or, in case an error occurs during reference resolution,
+     *         a {@link ReferenceResolutionException}.
+     */
+    @Nonnull
+    public CompletionStage<CustomerDraftBuilder> resolveCustomerGroupReference(
+        @Nonnull final CustomerDraftBuilder draftBuilder) {
+
+        final ResourceIdentifier<CustomerGroup> customerGroupResourceIdentifier = draftBuilder.getCustomerGroup();
+        if (customerGroupResourceIdentifier != null && customerGroupResourceIdentifier.getId() == null) {
+            String customerGroupKey;
+            try {
+                customerGroupKey = getKeyFromResourceIdentifier(customerGroupResourceIdentifier);
+            } catch (ReferenceResolutionException referenceResolutionException) {
+                return exceptionallyCompletedFuture(new ReferenceResolutionException(
+                    format(FAILED_TO_RESOLVE_CUSTOMER_GROUP_REFERENCE, draftBuilder.getKey(),
+                        referenceResolutionException.getMessage())));
+            }
+
+            return fetchAndResolveCustomerGroupReference(draftBuilder, customerGroupKey);
+        }
+        return completedFuture(draftBuilder);
+    }
+
+    @Nonnull
+    private CompletionStage<CustomerDraftBuilder> fetchAndResolveCustomerGroupReference(
+        @Nonnull final CustomerDraftBuilder draftBuilder,
+        @Nonnull final String customerGroupKey) {
+
+        return customerGroupService
+            .fetchCachedCustomerGroupId(customerGroupKey)
+            .thenCompose(resolvedCustomerGroupIdOptional -> resolvedCustomerGroupIdOptional
+                .map(resolvedCustomerGroupId ->
+                    completedFuture(draftBuilder.customerGroup(
+                        CustomerGroup.referenceOfId(resolvedCustomerGroupId).toResourceIdentifier())))
+                .orElseGet(() -> {
+                    final String errorMessage = format(CUSTOMER_GROUP_DOES_NOT_EXIST, customerGroupKey);
+                    return exceptionallyCompletedFuture(new ReferenceResolutionException(
+                        format(FAILED_TO_RESOLVE_CUSTOMER_GROUP_REFERENCE, draftBuilder.getKey(),
+                            errorMessage)));
+                }));
+    }
+
+    /**
+     * Given a {@link CustomerDraftBuilder} this method attempts to resolve the stores and return
+     * a {@link CompletionStage} which contains a new instance of the builder with the resolved references.
+     *
+     * @param draftBuilder the customer draft to resolve its store references.
+     * @return a {@link CompletionStage} that contains as a result a new builder instance with resolved references or,
+     *         in case an error occurs during reference resolution, a {@link ReferenceResolutionException}.
+     */
+    @Nonnull
+    public CompletionStage<CustomerDraftBuilder> resolveStoreReferences(
```

Comment:
I don't want to handle the checked exception here, so the stage is not complete for the exceptional case.

Reply:
I see, but if you throw here instead of doing `exceptionallyCompletedFuture`, will it be caught anyway in a `.exceptionally` clause or you need to explicitly do try/catch? I'm not sure how it works in Java, but I thought it's the same way as in JS - if you have a an exception thrown in a Promise, it will be caught also in a Promise. 

```
    public CustomerDraftBuilder resolveStoreReferences(
        @Nonnull final CustomerDraftBuilder draftBuilder) throws ReferenceResolutionException {

        final List<ResourceIdentifier<Store>> storeResourceIdentifiers = draftBuilder.getStores();
        if (storeResourceIdentifiers == null || storeResourceIdentifiers.isEmpty()) {
            return draftBuilder;
        }
        final List<ResourceIdentifier<Store>> resolvedReferences = new ArrayList<>();
        for (ResourceIdentifier<Store> storeResourceIdentifier : storeResourceIdentifiers) {
            if (storeResourceIdentifier != null) {
                if (storeResourceIdentifier.getId() == null) {
                    try {
                        final String storeKey = getKeyFromResourceIdentifier(storeResourceIdentifier);
                        resolvedReferences.add(ResourceIdentifier.ofKey(storeKey));
                    } catch (ReferenceResolutionException referenceResolutionException) {
                            throw new ReferenceResolutionException(
                                format(FAILED_TO_RESOLVE_STORE_REFERENCE,
                                    draftBuilder.getKey(), referenceResolutionException.getMessage()));
                    }
                } else {
                    resolvedReferences.add(ResourceIdentifier.ofId(storeResourceIdentifier.getId()));
                }
            }
        }
        return draftBuilder.stores(resolvedReferences);
    }
```

Reply:
You are not wrong with saying if it's not an async operation why you are using it, here it's my preference to make less effort, if I throw a checked exception here other methods need to implement it, which will go till the sync implementation so here exception handling is using the futures, If it's ok for you, I want to keep it as is.